sdks/python/apache_beam/io/filebasedsource.py

```diff
@@ -0,0 +1,246 @@
+          start_position, stop_position)
+
+    for source in self._sources:
+      # We assume all sub-sources to produce bundles that specify weight using
+      # the same unit. For example, all sub-sources may specify the size in
+      # bytes as their weight.
+      for bundle in source.split(desired_bundle_size, None, None):
+        yield bundle
+
+  def get_range_tracker(self, start_position, stop_position):
+    assert start_position is None
+    assert stop_position is None
+    # This will be invoked only when FileBasedSource is read without splitting.
+    # For that case, we only support reading the whole source.
+    return range_trackers.OffsetRangeTracker(0, len(self.sources))
+
+  def read(self, range_tracker):
+    for index, sub_source in enumerate(self.sources):
+      if not range_tracker.try_claim(index):
+        return
+
+      sub_source_tracker = sub_source.get_range_tracker(None, None)
+      for record in sub_source.read(sub_source_tracker):
+        yield record
+
+  def default_output_coder(self):
+    if self._sources:
+      # Getting coder from the first sub-sources. This assumes all sub-sources
+      # to produce the same coder.
+      return self._sources[0].default_output_coder()
+    else:
+      # Defaulting to PickleCoder.
+      return super(_ConcatSource, self).default_output_coder()
+
+
+class FileBasedSource(iobase.BoundedSource):
+  """A ``BoundedSource`` for reading a file glob of a given type."""
+
+  def __init__(self, file_pattern, min_bundle_size=0):
+    """Initializes ``FileBasedSource``.
+
+    Args:
+      file_pattern: the file glob to read.
+      min_bundle_size: minimum size of bundles that should be generated when
+                       performing initial splitting on this source.
+    """
+    self._pattern = file_pattern
+    self._concat_source = None
+    self._min_bundle_size = min_bundle_size
+
+  def _get_concat_source(self):
+    if self._concat_source is None:
+      single_file_sources = []
+      file_names = [f for f in fileio.ChannelFactory.glob(self._pattern)]
+      sizes = FileBasedSource._estimate_sizes_in_parallel(file_names)
+
+      for index, file_name in enumerate(file_names):
+        if sizes[index] == 0:
+          continue  # Ignoring empty file.
+
+        single_file_source = _SingleFileSource(
+            self, file_name,
+            0,
+            sizes[index],
+            min_bundle_size=self._min_bundle_size)
+        single_file_sources.append(single_file_source)
+      self._concat_source = _ConcatSource(single_file_sources)
+    return self._concat_source
+
+  def open_file(self, file_name):
+    return fileio.ChannelFactory.open(
+        file_name, 'rb', 'application/octet-stream')
+
+  @staticmethod
+  def _estimate_sizes_in_parallel(file_names):
+
+    def _calculate_size_of_file(file_name):
+      f = fileio.ChannelFactory.open(
+          file_name, 'rb', 'application/octet-stream')
+      try:
+        f.seek(0, os.SEEK_END)
```

Comment:
Yeah, that's a better approach.

Added a method to 'ChannelFactory' and updated 'filebasedsource' to use that.